firestore/integration_test_internal/src/bundle_test.cc

```diff
@@ -0,0 +1,196 @@
+#include "firestore_integration_test.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "util/bundle_builder.h"
+#include "util/event_accumulator.h"
+
+// These test cases are in sync with native iOS client SDK test
+//   Firestore/Example/Tests/Integration/API/FIRBundlesTests.mm
+// and native Android client SDK test
+//   firebase_firestore/tests/integration_tests/src/com/google/firebase/firestore/BundleTest.java
+// The iOS test names start with the mandatory test prefix while Android test
+// names do not. Here we use the Android test names.
+
+namespace firebase {
+namespace firestore {
+namespace {
+
+void VerifyProgress(LoadBundleTaskProgress progress, int expected_documents) {
+  EXPECT_EQ(progress.state(), LoadBundleTaskProgress::State::kInProgress);
+  EXPECT_EQ(progress.documents_loaded(), expected_documents);
+  EXPECT_LE(progress.documents_loaded(), progress.total_documents());
+  EXPECT_LE(progress.bytes_loaded(), progress.total_bytes());
+}
+
+void VerifySuccessProgress(LoadBundleTaskProgress progress) {
+  EXPECT_EQ(progress.state(), LoadBundleTaskProgress::State::kSuccess);
+  EXPECT_EQ(progress.documents_loaded(), progress.total_documents());
+  EXPECT_EQ(progress.bytes_loaded(), progress.total_bytes());
+}
+
+void VerifyErrorProgress(LoadBundleTaskProgress progress) {
+  EXPECT_EQ(progress.state(), LoadBundleTaskProgress::State::kError);
+  EXPECT_EQ(progress.documents_loaded(), 0);
+  EXPECT_EQ(progress.bytes_loaded(), 0);
+}
+
+class BundleTest : public FirestoreIntegrationTest {
+ protected:
+  void SetUp() override { Await(TestFirestore()->ClearPersistence()); }
+
+  void VerifyQueryResults(Firestore *db) {
+    {
+      auto snapshot = *Await(db->Collection("coll-1").Get(Source::kCache));
+      EXPECT_THAT(
+          QuerySnapshotToValues(snapshot),
+          testing::ElementsAre(MapFieldValue{{"k", FieldValue::String("a")},
+                                             {"bar", FieldValue::Integer(1)}},
+                               MapFieldValue{{"k", FieldValue::String("b")},
+                                             {"bar", FieldValue::Integer(2)}}));
+    }
+
+    {
+      Query limit = *Await(db->NamedQuery("limit"));
+      auto limit_snapshot = *Await(limit.Get(Source::kCache));
+      EXPECT_THAT(
+          QuerySnapshotToValues(limit_snapshot),
+          testing::ElementsAre(MapFieldValue{{"k", FieldValue::String("b")},
+                                             {"bar", FieldValue::Integer(2)}}));
+    }
+
+    {
+      Query limit_to_last = *Await(db->NamedQuery("limit-to-last"));
+      auto limit_to_last_snapshot = *Await(limit_to_last.Get(Source::kCache));
+      EXPECT_THAT(
+          QuerySnapshotToValues(limit_to_last_snapshot),
+          testing::ElementsAre(MapFieldValue{{"k", FieldValue::String("a")},
+                                             {"bar", FieldValue::Integer(1)}}));
+    }
+  }
+};
+
+TEST_F(BundleTest, CanLoadBundlesWithProgressUpdates) {
```

Comment:
Please add a test that explicitly tests loading bundles *without* progress updates.